### DISCUSSION TITLE
Issues to execute queries via SQL Lab

### DISCUSSION THREAD

**Discussion by sprudel79** [apache/superset#26784] [Q&A / Help]
Hi everyone,
I have a K8s cluster where I have installed the latest helm chart of superset from http://apache.github.io/superset/ (which is 3.1.0).
I need to access several MS SQL databases in my dashboards. For the creation of the database connections I use this kind of syntax for the SQLAlchemy uri:  "mssql+pymssql://<user>:<password>@<server>:<port>/<database>
When clicking the "Test connection" button, I get informed that everything's fine.
Now to the issue:
Next thing is to go to the SQL Lab to verify the DB connections. I can browse the available ones in the "DATABASE" drop-down, I am able to see the correct "SCHEMA" options as well as the "TABLE SCHEMA" options. SQL Lab shows also nicely all the available columns, so in my opinion the DB connection settings should be fine.
However, when I try to execute a query (I tried even with a super simple one like "SELECT 1 AS A") I always get the error:

**Microsoft SQL Server Error
Authentication required.
This may be triggered by:
Issue 1002 - The database returned an unexpected error.** 

In the logs of the superset pod I can see this:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/usr/local/lib/python3.9/site-packages/flask_appbuilder/security/decorators.py", line 95, in wraps
    return f(self, *args, **kwargs)
  File "/app/superset/views/base_api.py", line 127, in wraps
    raise ex
  File "/app/superset/views/base_api.py", line 121, in wraps
    duration, response = time_function(f, self, *args, **kwargs)
  File "/app/superset/utils/core.py", line 1454, in time_function
    response = func(*args, **kwargs)
  File "/app/superset/views/base_api.py", line 93, in wraps
    return f(self, *args, **kwargs)
  File "/app/superset/utils/log.py", line 255, in wrapper
    value = f(*args, **kwargs)
  File "/app/superset/sqllab/api.py", line 407, in execute_sql_query
    command_result: CommandResult = command.run()
  File "/app/superset/commands/sql_lab/execute.py", line 121, in run
    raise ex
  File "/app/superset/commands/sql_lab/execute.py", line 103, in run
    status = self._run_sql_json_exec_from_scratch()
  File "/app/superset/commands/sql_lab/execute.py", line 161, in _run_sql_json_exec_from_scratch
    raise ex
  File "/app/superset/commands/sql_lab/execute.py", line 156, in _run_sql_json_exec_from_scratch
    return self._sql_json_executor.execute(
  File "/app/superset/sqllab/sql_json_executer.py", line 111, in execute
    raise SupersetErrorsException(
superset.exceptions.SupersetErrorsException: [SupersetError(message='Authentication required.', error_type=<SupersetErrorType.GENERIC_DB_ENGINE_ERROR: 'GENERIC_DB_ENGINE_ERROR'>, level=<ErrorLevel.ERROR: 'error'>, extra={'engine_name': 'Microsoft SQL Server', 'issue_codes': [{'code': 1002, 'message': 'Issue 1002 - The database returned an unexpected error.'}]})]
```

But now again another weird thing is that I can use the same DB connection without any problem in a chart, so it seems only SQL Lab has issues. Is there anything I did wrong?

I am aware that the error is described over here: 
https://superset.apache.org/docs/miscellaneous/issue-codes/#issue-1002

But I have my doubts there's a problem with the DB itself since the connection works in general and even in the charts.

Many thanks in advance and kind regards,
Sprudel

**Comment by craunas** [comment]
@sprudel79 Did you find any way around this?

We're experiencing the same problem but with Snowflake as data source.
Looks like the query gets down to the DB but no results are shown in sql lab.

We've got the same error message as in the original post but have gotten this one as well:
`Protocol error: unauthenticated bulk length`

which seem to be related to Redis?

**Reply by sprudel79** [reply]
Hi @craunas , unfortunately I haven't found any solution to this issue. It cannot be a general issue with the connection itself because I can use it in SQL Lab and browse the tables and columns. Only when executing the query I get the error. Next step for me would be to check the source code directly.

**Reply by craunas** [reply]
Hi! 
Sounds exactly like the problem we had. We solved it by using an in-cluster Redis setup instead of a Azure managed one.
Not sure why it solved it but one of our devops guys thought it would be because Azure only offers version 6 of Redis while the one we deployed in cluster ran version 6.

Let me know if I can help out in any way @sprudel79

**Reply by sprudel79** [reply]
Thanks @craunas , indeed we've been using an external Redis instance in our setup instead the default in-cluster one. So as I understand I should simply try the in-cluster setup to solve this issue?

**Reply by craunas** [reply]
It solved our issues at least, so it could be worth a try at least.
In our case we decided that since its only used for caching its not as critical as the backend db's fore example, so no risk having it in-cluster instead. @sprudel79

**Comment by Samoth69** [comment]
Make sure your `RESULT_BACKEND` or RedisCache config in superset_config.py is working. In our case we missed a [password field in RedisCache object](https://github.com/pallets-eco/flask-caching/blob/master/src/flask_caching/backends/rediscache.py#L31).

**Reply by dimashenkov** [reply]
and you had `message='Authentication required.', error_type=<SupersetErrorType.GENERIC_DB_ENGINE_ERROR` ?

**Reply by sprudel79** [reply]
Thanks @dimashenkov for asking.
I indeed get this error:

![image](https://github.com/user-attachments/assets/77586552-9d89-4b32-b87b-bb9a02ed99dd)

I have tried already a few things regarding Redis (avoid using external Redis and rather using the instance which will get deployed automatically along with the Helm chart as it was suggested in this thread) but no luck so far.
We have considered removing the SQLLab from the menu which as I understand is a matter of setting the right permissions.

**Reply by dimashenkov** [reply]
@sprudel79 I fixed the ERR with this change in helpers.tpl , its a bug and I would create an issue from that discussion .
```
CELERY_CONFIG = CeleryConfig
RESULTS_BACKEND = RedisCache(
      host=env('REDIS_HOST'),
{{- if or .Values.redis.auth.enabled .Values.supersetNode.connections.redis_password }}
      password=env('REDIS_PASSWORD'),
{{- end }}
      port=env('REDIS_PORT'),
```

**Reply by sprudel79** [reply]
Hi @dimashenkov , awesome, many thanks for having shared that fix!
I can confirm it's working on my side, I have simply applied the following code in my configOverrides section at the very end:

```
RESULTS_BACKEND = RedisCache(
      host=env('REDIS_HOST'),
      password=env('REDIS_PASSWORD'),
      port=env('REDIS_PORT'),
      key_prefix='superset_results',      
    )
```

And yes, that did the trick! Once you have the issue created we can close this one. 
Thanks again!

**Comment by shermanLF** [comment]
A misconfigured RESULT_CACHE fails in a non-obvious way. There are no exceptions logged related to the RESULT_CACHE.

The fix is to have a working RESULT_CACHE.

Superset needs to do a better job indicating there are problems with the RESULT_CACHE and other caches.

**My journey**

**The Superset environment**
- Superset in an ECS task
- AWS Redis Elasticache
- AWS PostgreSQL for Superset database and data sources.

**Set up**
- The connection of Superset to its own database working, which uses SQLAlchemy like SQL Lab.
- Creating the database connections through the Superset UI works - either with the dialog or manually putting in a SQLAlchemy URI.
- Connections pass validation.
- Charts in Superset can be created based on the data source.

**SQL Lab problem**
- Browsing the metadata of a database works.
- Running a query in the Lab fails.

**Error**

```
PostgreSQL Error
Invalid input of type: 'set'. Convert to a bytes, string, int or float first.
```

From the activity log:
`{"source": "sqlLab", "source_id": "6", "db_id": 5, "schema": "public", "ts": 1746105365220, "event_name": "sqllab_fetch_failed_query", "has_err": true, "start_offset": 1746105364455, "error_type": "GENERIC_DB_ENGINE_ERROR", "error_details": "Issue 1002 - The database returned an unexpected error.", "event_type": "timing", "trigger_event": "L-QKeFkiCxhWmH2RF5xg9"}`

**Result**
Of course this looks like a database related issue :-( 
Queue days of database log and configuration analysis.

I changed the RESULT_CACHE to use a database backend rather than Redis Elasticache, and the SQL Lab error disappears.
